plugins/pixart-rf/fu-pxi-ble-device.c

```diff
@@ -91,33 +91,52 @@ fu_pxi_ble_device_prepare_firmware(FuDevice *device,
 				   FwupdInstallFlags flags,
 				   GError **error)
 {
-	FuPxiBleDevice *self = FU_PXI_BLE_DEVICE(device);
-	const gchar *model_name;
 	g_autoptr(FuFirmware) firmware = fu_pxi_firmware_new();
 
 	if (!fu_firmware_parse(firmware, fw, flags, error))
 		return NULL;
 
-	/* check is compatible with hardware */
-	model_name = fu_pxi_firmware_get_model_name(FU_PXI_FIRMWARE(firmware));
-	if ((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
-		if (self->model_name == NULL || model_name == NULL) {
-			g_set_error_literal(error,
+	if (fu_device_has_private_flag(device, FU_PXI_DEVICE_FLAG_IS_HPAC) &&
+	    fu_pxi_firmware_get_is_hpac_flag(FU_PXI_FIRMWARE(firmware))) {
+		g_autoptr(GBytes) fw_tmp = NULL;
+		gsize hpac_fw_size = 0;
+		const guint8 *fw_ptr = g_bytes_get_data(fw,NULL);
+
+		hpac_fw_size = (fw_ptr[9] | (fw_ptr[10] << 8) | (fw_ptr[11] << 16) | (fw_ptr[12] << 24)) + 264;
+		fw_tmp = g_bytes_new_from_bytes(fw,9,hpac_fw_size);
+		fu_firmware_set_bytes(firmware,fw_tmp);
+	} else if (!fu_device_has_private_flag(device, FU_PXI_DEVICE_FLAG_IS_HPAC) &&
+		   !fu_pxi_firmware_get_is_hpac_flag(FU_PXI_FIRMWARE(firmware))) {
+		FuPxiBleDevice *self = FU_PXI_BLE_DEVICE(device);
+		const gchar *model_name;
+
+		/* check is compatible with hardware */
+		model_name = fu_pxi_firmware_get_model_name(FU_PXI_FIRMWARE(firmware));
+		if ((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
+			if (self->model_name == NULL || model_name == NULL) {
+				g_set_error_literal(error,
+						    FWUPD_ERROR,
+						    FWUPD_ERROR_INVALID_FILE,
+						    "legacy device or firmware detected, "
+						    "--force required");
+				return NULL;
+			}
+			if (g_strcmp0(self->model_name, model_name) != 0) {
+				g_set_error(error,
 					    FWUPD_ERROR,
 					    FWUPD_ERROR_INVALID_FILE,
-					    "legacy device or firmware detected, "
-					    "--force required");
-			return NULL;
-		}
-		if (g_strcmp0(self->model_name, model_name) != 0) {
-			g_set_error(error,
-				    FWUPD_ERROR,
-				    FWUPD_ERROR_INVALID_FILE,
-				    "incompatible firmware, got %s, expected %s.",
-				    model_name,
-				    self->model_name);
-			return NULL;
+					    "incompatible firmware, got %s, expected %s.",
+					    model_name,
+					    self->model_name);
+				return NULL;
+			}
 		}
+	} else {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "The firmware is incompatible with quirk flag");
```

Comment:
Although a quirk can override it; I think it's really a device not a quirk problem.
```suggestion
			    "The firmware is incompatible with this device");
```

Reply:
revised in d60a6699e24f22fbb4b95a07d466b26a0ab63d3a